lib/std/fs.zig

```diff
@@ -582,10 +582,33 @@ pub const Dir = struct {
             const path_w = try os.windows.sliceToPrefixedFileW(sub_path);
             return self.openFileW(&path_w, flags);
         }
+        if (builtin.os.tag == .wasi) {
+            return self.openFileWasi(sub_path, flags);
+        }
         const path_c = try os.toPosixPath(sub_path);
         return self.openFileZ(&path_c, flags);
     }
 
+    fn openFileWasi(self: Dir, sub_path: []const u8, flags: File.OpenFlags) File.OpenError!File {
+        var fdflags: wasi.fdflag_t = 0x0;
+        var rights: wasi.rights_t = 0x0;
+        if (flags.read) {
+            rights |= wasi.FD_READ | wasi.FD_TELL | wasi.FD_FILESTAT_GET;
+        }
+        if (flags.write) {
+            fdflags |= wasi.FDFLAG_APPEND;
+            rights |= wasi.FD_WRITE | wasi.FD_DATASYNC | wasi.FD_SEEK | wasi.FD_FDSTAT_SET_FLAGS | wasi.FD_SYNC | wasi.FD_ALLOCATE | wasi.FD_ADVISE | wasi.FD_FILESTAT_SET_TIMES | wasi.FD_FILESTAT_SET_SIZE;
+        }
+
+        const fd = try os.openatWasi(self.fd, sub_path, 0x0, fdflags, rights);
+
+        return File{
+            .handle = fd,
+            .io_mode = .blocking,
+            .async_block_allowed = File.async_block_allowed_no,
```

Comment:
I would recommend to rebase this branch in light of recent changes to async I/O. For example these two fields have been renamed to `capable_io_mode` and `intended_io_mode` respectively. Also they have default initialization values which I believe are what you should use on WASI. So this could be changed to simply `File{.handle = fd}`. Does WASI support non-blocking I/O?

Reply:
Cool, thanks for the pointer!

Right, so we want WASI to offer non-blocking I/O but we're not there yet. For WASI to offer non-blocking I/O, we need changes to Wasm spec itself (multi-threading) and those haven't landed yet. So for the moment, I suggest we draft it out specifying that you can only issue blocking WASI calls. Oh, and just as an FYI, there is a non-blocking fdflag in the WASI [`wasi.FDFLAG_NONBLOCK`](https://github.com/ziglang/zig/blob/master/lib/std/os/bits/wasi.zig#L151) that we'll want to set when requesting non-blocking I/O, but currently this flag doesn't do much, except being translated into an appropriate non-blocking equivalent flag on the host (such as `O_NONBLOCK` in posix, or `FILE_FLAG_OVERLAPPED` on Windows).

Reply:
That's good to hear that it is planned. I don't think you need multi-threading to accomplish nonblocking I/O. Node.js for example is single threaded non-blocking I/O.